Add checks for the supported WebCrypto algorithms.

javifernandez opened the pull request:
#### 1fa83f2bc4ee72f8ebf13b4dae1348c2f4d1897a
<pre>
Add checks for the supported WebCrypto algorithms.
<a href="https://bugs.webkit.org/show_bug.cgi?id=261550">https://bugs.webkit.org/show_bug.cgi?id=261550</a>

Reviewed by Žan Doberšek.

We have several functions using if/then/else clauses and ternary
operators to  reuse the logic structure between several algorithms
that use the OKP keys (eg, Curve25519 and Curve488).

While it&apos;s not a big deal for Curve25519, since we are talking just
about 2 algorithms, it may become a problem if we want to add support
for Curve488 in the future.

This PR adds some checks in different functions to early return in
case of unsupported algorithms.

* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::algorithmId):
(WebCore::CryptoKeyOKP::exportSpki const):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::exportPkcs8 const):
* Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::writeOID):
(WebCore::CryptoKeyOKP::importPkcs8):

Canonical link: <a href="https://commits.webkit.org/268079@main">https://commits.webkit.org/268079@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f28e6b0e48678bb51f3147171bf404ab7e1baa35

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18401 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18739 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19332 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20250 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17228 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18596 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22045 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18888 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19133 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18625 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18817 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16026 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21129 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16040 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16783 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23262 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17060 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16953 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21149 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17528 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14871 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16609 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4423 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20976 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17367 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -205,7 +215,18 @@ RefPtr<CryptoKeyOKP> CryptoKeyOKP::importSpki(CryptoAlgorithmIdentifier identifi
 
         // Construct the `public-key` expression to be used for generating the MPI structure.
-        gcry_error_t error = gcry_sexp_build(&platformKey, nullptr, curve == CryptoKeyOKP::NamedCurve::Ed25519 ? "(public-key(ecc(curve Ed25519)(q %b)))" : "(public-key(ecc(curve Curve25519)(q %b)))", subjectPublicKey->size(), subjectPublicKey->data());
+        gcry_error_t error;
```
This should still be initialized, to GPG_ERR_NO_ERROR.

javifernandez replied:
done

zdobersek commented on Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
```diff
@@ -272,6 +289,9 @@ RefPtr<CryptoKeyOKP> CryptoKeyOKP::importPkcs8(CryptoAlgorithmIdentifier identif
             break;
+        default:
+            ASSERT_NOT_REACHED();
+            break;
```
This case should also avoid further execution and return nullptr.

javifernandez replied:
The function returns void and there is no further code to execute after this, so maybe not needed ?

zdobersek replied:
Unless the GitHub UI is broken, I think this one returns `RefPtr<CryptoKeyOKP>`, so it should return nullptr. But `writeOID()` above does return void and there's nothing else after that switch statement, so a break there is fine.

javifernandez replied:
oh, true, I totally misread it :( 
Will fix it now.

zdobersek commented on Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
```diff
@@ -44,28 +44,33 @@ bool CryptoKeyOKP::isPlatformSupportedCurve(NamedCurve namedCurve)
-
-    if (identifier == CryptoAlgorithmIdentifier::Ed25519) {
-        int ccerror = 0;
+    int ccerror = 0;
```
Nit: `ccerror` doesn't really do anything outside of the two scopes, so it could be simply declared and used in both curve-specific scopes, IMO.

javifernandez replied:
Done

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -324,7 +361,18 @@ RefPtr<CryptoKeyOKP> CryptoKeyOKP::importPkcs8(CryptoAlgorithmIdentifier identif
 
         // Construct the `private-key` expression that will also be used for the EC context.
-        gcry_error_t error = gcry_sexp_build(&platformKey, nullptr, curve == CryptoKeyOKP::NamedCurve::Ed25519 ? "(private-key(ecc(curve Ed25519)(flags eddsa)(d %b)))" : "(private-key(ecc(curve Curve25519)(d %b)))", privateKey->size(), privateKey->data());
+        gcry_error_t error;
```
Should also be initialized to `GPG_ERR_NO_ERROR`.

javifernandez committed:
Add checks for the supported WebCrypto algorithms.
https://bugs.webkit.org/show_bug.cgi?id=261550

Reviewed by Žan Doberšek.

We have several functions using if/then/else clauses and ternary
operators to  reuse the logic structure between several algorithms
that use the OKP keys (eg, Curve25519 and Curve488).

While it's not a big deal for Curve25519, since we are talking just
about 2 algorithms, it may become a problem if we want to add support
for Curve488 in the future.

This PR adds some checks in different functions to early return in
case of unsupported algorithms.

* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::algorithmId):
(WebCore::CryptoKeyOKP::exportSpki const):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::exportPkcs8 const):
* Source/WebCore/crypto/mac/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::writeOID):
(WebCore::CryptoKeyOKP::importPkcs8):

Canonical link: https://commits.webkit.org/268079@main
